### PR TITLE
Improve code logic and add fetch genome by ensembl and assembly name

### PR DESCRIPTION
**Pull request by bilalebi**
### JIRA Ticket
https://www.ebi.ac.uk/panda/jira/browse/EA-1090

### Changes
Added the possibility to fetch genome by `ensembl_name` and `assembly_name` (`genebuild_id` will be added in the future)

---

Update:

> [Major changes](https://github.com/Ensembl/ensembl-metadata-api/pull/31#issuecomment-1651656468)
> * Added fetch genome by ensembl name and assembly name 
> * Remove mutual exclusivity in `fetch_genomes()` and `fetch_sequences()`
> * Improved `fetch_taxonomy_names()` to include `ncbi_common_name`, `alternative_names` and `common_name`
> * Improved `fetch_sequences()` to include `assembly_uuid`
> * Added `fetch_genome_by_keyword()`
> * Added and fixed broken tests
> * Improved the overall code logic
> * Added the possibility to fetch unreleased genomes
> * ...

This branch is also used to update the API to be used by [ensembl-metadata-service](https://github.com/Ensembl/ensembl-metadata-service)

### Related PR:
https://github.com/Ensembl/ensembl-metadata-service/pull/25

### CODE REVIEW

File: src/ensembl/production/metadata/api/genome.py [hunk 1]
```diff
@@ -80,9 +80,46 @@ def fetch_taxonomy_ids(self, taxonomy_names):
                 taxids.append(taxid[0])
         return taxids
 
-    def fetch_genomes(self, genome_id=None, genome_uuid=None, assembly_accession=None, ensembl_name=None,
-                      taxonomy_id=None, group=None, group_type=None, unreleased_only=False, site_name=None,
-                      release_type=None, release_version=None, current_only=True):
+    def fetch_genomes(self, genome_id=None, genome_uuid=None, assembly_accession=None, assembly_name=None,
+                      ensembl_name=None, genebuild_id=None, taxonomy_id=None, group=None, group_type=None,
```
Comment (dpopleton) [inline src/ensembl/production/metadata/api/genome.py:84]: genebuild_id is not implemented and should not be part of this PR.
This could be it's own method within the class as it isn't trivial to get. It will also only ever return one genome so it could be interesting.
Better might be dataset_name or other aspect of dataset. Those could be very useful.

File: src/ensembl/production/metadata/api/genome.py [hunk 2]
```diff
@@ -129,11 +169,18 @@ def fetch_genomes(self, genome_id=None, genome_uuid=None, assembly_accession=Non
...
         elif assembly_accession is not None:
             genome_select = genome_select.filter(Assembly.accession.in_(assembly_accession))
 
+        # Handle when both assembly_name and ensembl_name are provided
+        # Todo: add genebuild_id once it's in the metadata DB (EA-1090)
+        if assembly_name is not None and ensembl_name is not None:
+            genome_select = genome_select.filter(
+                db.and_(Assembly.name.in_([assembly_name]), Organism.ensembl_name.in_([ensembl_name]))
+            )
+
```
Comment (dpopleton) [inline src/ensembl/production/metadata/api/genome.py:183]: This is against the logic of the method.
Each separate parameter adds an individual filter which are combined at the end. 
Checking two at once is excessive and unnecessary. 
Adding assembly name is a good idea, but I would put it on its own.

**Review by dpopleton — CHANGES_REQUESTED** [review]

File: src/ensembl/production/metadata/api/genome.py [hunk 3]
```diff
@@ -104,11 +141,12 @@ def fetch_genomes(self, genome_id=None, genome_uuid=None, assembly_accession=Non
                 .join(OrganismGroupMember.organism_group) \
                 .filter(OrganismGroup.type.in_(group_type)).filter(OrganismGroup.name.in_(group))
 
+        # Apply additional filters based on the provided parameters
```
Comment (dpopleton) [inline src/ensembl/production/metadata/api/genome.py:144]: Have you tested these together? There are a few joins that might have to be done together before the filtering, but I am not sure.

File: src/tests/test_api.py [hunk 4]
```diff
@@ -82,6 +82,12 @@ def test_fetch_genomes_by_genome_uuid(self, multi_dbs):
         test = conn.fetch_genomes_by_genome_uuid('a733550b-93e7-11ec-a39d-005056b38ce3')
         assert test[0].Organism.scientific_name == 'Caenorhabditis elegans'
 
+    def test_fetch_genome_by_ensembl_and_assembly_name(self, multi_dbs):
+        conn = GenomeAdaptor(metadata_uri=multi_dbs['ensembl_metadata'].dbc.url,
+                             taxonomy_uri=multi_dbs['ncbi_taxonomy'].dbc.url)
+        test = conn.fetch_genomes(assembly_name='WBcel235', ensembl_name='caenorhabditis_elegans')
+        assert test[0].Organism.scientific_name == 'Caenorhabditis elegans'
+
     def test_fetch_genomes_by_assembly_accession(self, multi_dbs):
```
Comment (dpopleton) [inline src/tests/test_api.py:91]: Would you add a couple more tests to check your changes?

- everything is specified
- specifying with a few direct conflicts

**Review by dpopleton — CHANGES_REQUESTED** [review]

File: src/tests/test_api.py (hunk 4 above)
Comment (bilalebi) [reply src/tests/test_api.py:130]: Done!

File: src/ensembl/production/metadata/api/genome.py (hunk 3 above)
Comment (bilalebi) [reply src/ensembl/production/metadata/api/genome.py:144]: I added some tests and played with the code separately, I didn't get any unexpected behaviour

**Comment by bilalebi** [comment]
Hi @vinay-ebi 
Please feel free to take a look at this PR, some of the major changes:
* Added fetch genome by ensembl name and assembly name 
* Remove mutual exclusivity in `fetch_genomes()` and `fetch_sequences()`
* Improved `fetch_taxonomy_names()` to include `ncbi_common_name`, `alternative_names` and `common_name`
* Added `fetch_genome_by_keyword()`
* Added and fixed broken tests
* Added `__repr__` functions in the models
* Improved the overall code logic

> NB: I commented 3 out of 4 tests in the updater because they were failing and they need to be fixed based on the changes @dpopleton made in this PR https://github.com/Ensembl/ensembl-metadata-api/pull/32 that can be done later

File: src/ensembl/production/metadata/api/genome.py [hunk 5]
```diff
@@ -80,64 +112,118 @@ def fetch_taxonomy_ids(self, taxonomy_names):
...
-            ).join(Genome.assembly).join(Genome.organism) \
-                .join(Organism.organism_group_members) \
-                .join(OrganismGroupMember.organism_group) \
-                .filter(OrganismGroup.type.in_(group_type)).filter(OrganismGroup.name.in_(group))
+            genome_select = genome_select.filter(OrganismGroup.type.in_(group_type)).filter(OrganismGroup.name.in_(group))
 
+        # Apply additional filters based on the provided parameters
         if unreleased_only:
+            # Question (Bilal): When adding unreleased genomes do we populate genome_releases table?
+            # if that's the case the test data doesn't do so (this will return en empty list)
```
Comment (dpopleton) [inline src/ensembl/production/metadata/api/genome.py:183]: Unreleased genomes will not have a genome_release. Your logic is odd in my mind as I would look for genomes without a genome_release, but if it works I am fine with it.

File: src/ensembl/production/metadata/api/genome.py [hunk 6]
```diff
@@ -80,64 +112,118 @@ def fetch_taxonomy_ids(self, taxonomy_names):
...
+            genome_select = genome_select.filter(EnsemblSite.name == site_name)
 
             if release_type is not None:
                 genome_select = genome_select.filter(EnsemblRelease.release_type == release_type)
 
             if current_only:
                 genome_select = genome_select.filter(GenomeRelease.is_current == 1)
 
-            if release_version is not None:
+            if release_version != 0.0:
```
Comment (dpopleton) [inline src/ensembl/production/metadata/api/genome.py:203]: So when we are releasaing a genome we will create the genome_release and link the genome_datasets. Until then there is nothing for those genomes.
The 0.0 here is not proper for any aspect of the API

File: src/ensembl/production/metadata/api/genome.py [hunk 7]
```diff
@@ -208,31 +294,94 @@ def fetch_genomes_by_scientific_name(
...
+                                                         Organism.scientific_parlance_name) == keyword.lower(),
+                                                     db.func.lower(Organism.species_taxonomy_id) == keyword.lower()))
+
+        if release_version == 0 or release_version is None:
+            genome_query = genome_query.where(EnsemblRelease.is_current == 1)
+        else:
+            genome_query = genome_query.where(EnsemblRelease.version <= release_version)
+
+        with self.metadata_db.session_scope() as session:
+            session.expire_on_commit = False
```
Comment (dpopleton) [inline src/ensembl/production/metadata/api/genome.py:334]: Make sure you need the session to keep in memory here. If so, fine.

File: src/ensembl/production/metadata/api/genome.py [hunk 8]
```diff
@@ -249,48 +398,106 @@ def fetch_sequences_by_assembly_accession(
...
 
+            if organism_uuid is not None:
+                genome_select = genome_select.filter(Organism.organism_uuid.in_(organism_uuid))
+
             if dataset_uuid is not None:
                 genome_select = genome_select.filter(Dataset.dataset_uuid.in_(dataset_uuid))
 
             if unreleased_datasets:
                 genome_select = genome_select.filter(GenomeDataset.release_id.is_(None)) \
                     .filter(GenomeDataset.is_current == 0)
```
Comment (dpopleton) [inline src/ensembl/production/metadata/api/genome.py:480]: Don't Filter off of is_current for unreleased. Old datasets will be present. Filter off the absence of a release.

File: src/ensembl/production/metadata/api/genome.py [hunk 9]
```diff
@@ -249,48 +398,106 @@ def fetch_sequences_by_assembly_accession(
...
-            if dataset_name is not None:
+
+            if is_released:
+                # Include release related info only if is_released is True
+                genome_select = genome_select.add_columns(EnsemblRelease, DatasetAttribute, Attribute) \
+                    .join(EnsemblRelease, GenomeDataset.release_id == EnsemblRelease.release_id) \
+                    .join(DatasetAttribute, DatasetAttribute.dataset_id == Dataset.dataset_id) \
+                    .join(Attribute, Attribute.attribute_id == DatasetAttribute.attribute_id)
+
+            if dataset_name is not None and "all" not in dataset_name:
```
Comment (dpopleton) [inline src/ensembl/production/metadata/api/genome.py:489]: Why is the "all" here?

File: src/ensembl/production/metadata/api/models/genome.py [hunk 10]
```diff
@@ -36,6 +36,9 @@ class Genome(Base):
     # organism_id to organism
     organism = relationship("Organism", back_populates="genomes")
 
+    def __repr__(self):
```
Comment (dpopleton) [inline src/ensembl/production/metadata/api/models/genome.py:39]: If you are going to add the repr, please be standard between all model classes. These are of a different format than assembly.

**Review by dpopleton — CHANGES_REQUESTED** [review]
So a few minor comments. It looks good for the most part. I have not tested it further or checked for conflicts , so I hope you have. :)

File: src/ensembl/production/metadata/api/genome.py (hunk 6 above)
Comment (bilalebi) [reply src/ensembl/production/metadata/api/genome.py]: done

File: src/ensembl/production/metadata/api/models/genome.py (hunk 10 above)
Comment (bilalebi) [reply src/ensembl/production/metadata/api/models/genome.py:39]: I removed them for now

File: src/ensembl/production/metadata/api/genome.py (hunk 9 above)
Comment (bilalebi) [reply src/ensembl/production/metadata/api/genome.py:475]: By default the API `dataset_name="assembly"`; `all` allows us to fetch everything
Please take a look at this discussion for more context: https://github.com/Ensembl/ensembl-metadata-api/commit/a8eb73d7e3d61d84510dc957cf87777f93e4434c#r121262595

File: src/ensembl/production/metadata/api/genome.py [hunk 11]
```diff
@@ -249,48 +398,106 @@ def fetch_sequences_by_assembly_accession(
...
                 genome_select = genome_select.filter(Genome.genome_uuid.in_(genome_uuid))
 
+            if organism_uuid is not None:
+                genome_select = genome_select.filter(Organism.organism_uuid.in_(organism_uuid))
+
             if dataset_uuid is not None:
                 genome_select = genome_select.filter(Dataset.dataset_uuid.in_(dataset_uuid))
 
             if unreleased_datasets:
                 genome_select = genome_select.filter(GenomeDataset.release_id.is_(None)) \
```
[the remark this answers is not in the document]
Comment (bilalebi) [reply src/ensembl/production/metadata/api/genome.py]: Done

File: src/ensembl/production/metadata/api/genome.py (hunk 8 above)
Comment (bilalebi) [reply src/ensembl/production/metadata/api/genome.py]: fixed

File: src/ensembl/production/metadata/api/genome.py (hunk 9 above)
Comment (dpopleton) [reply src/ensembl/production/metadata/api/genome.py:475]: Ok, if it works, I am happy.

**Review by dpopleton — DISMISSED** [review]

File: src/ensembl/production/metadata/api/genome.py [hunk 12]
```diff
@@ -201,31 +273,94 @@ def fetch_genomes_by_scientific_name(
...
+        """
         genome_id = check_parameter(genome_id)
         genome_uuid = check_parameter(genome_uuid)
+        assembly_uuid = check_parameter(assembly_uuid)
         assembly_accession = check_parameter(assembly_accession)
+        assembly_sequence_accession = check_parameter(assembly_sequence_accession)
 
-        seq_select = db.select(AssemblySequence, )
+        seq_select = db.select(
+            Genome, Assembly, AssemblySequence
```
[the remark this answers is not in the document]
Comment (bilalebi) [reply src/ensembl/production/metadata/api/genome.py:339]: Yes, they're used in the Metadata service to fetch data (e.g. `accession`, `level`...)

File: src/ensembl/production/metadata/api/genome.py (hunk 7 above)
Comment (bilalebi) [reply src/ensembl/production/metadata/api/genome.py:313]: Yes, it's needed.
When commenting it some tests start throwing this error:
```
sqlalchemy.orm.exc.DetachedInstanceError: Instance <DatasetType at 0x7fc5c05a13d0> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: https://sqlalche.me/e/14/bhk3)
```

File: src/ensembl/production/metadata/api/genome.py [hunk 13]
```diff
@@ -73,66 +74,137 @@ def fetch_taxonomy_ids(self, taxonomy_names):
...
+        if ensembl_name is not None:
             genome_select = genome_select.filter(Organism.ensembl_name.in_(ensembl_name))
 
-        elif taxonomy_id is not None:
+        if taxonomy_id is not None:
             genome_select = genome_select.filter(Organism.taxonomy_id.in_(taxonomy_id))
 
+        # Check if genome is released
+        with self.metadata_db.session_scope() as session:
+            session.expire_on_commit = False
```
[the remark this answers is not in the document]
Comment (bilalebi) [reply src/ensembl/production/metadata/api/genome.py:173]: Fixed

**Review by marcoooo — APPROVED** [review]
